### PR TITLE
fix(sidebar): bubble hidden lineage stream state

### PR DESCRIPTION
**Pull request by santastabber**
## Thinking Path

- Sidebar rows can collapse multiple WebUI continuation/lineage sessions into one visible conversation.
- A running continuation can be hidden from the normal sidebar list when it is archived, while the user-facing conversation remains visible through another lineage row.
- The visible row still needs to show running/completion state, otherwise users lose the spinner and activity ordering for the chat they interacted with.
- The fix keeps hidden archived lineage rows hidden, but bubbles their runtime/sidebar state onto the visible lineage row.
- Pin/unpin uses the same sidebar cache, so it should update the cached row and re-render immediately before the slower server reconciliation.

## What Changed

- Treat server-confirmed `is_streaming`, explicit pending-message markers, and active local `S.busy` as effective sidebar streaming state, without trusting stale raw `active_stream_id` or `pending_started_at` values by themselves.
- Bubble hidden archived lineage child state to the visible lineage row:
  - running spinner state,
  - latest activity timestamp for ordering/date grouping,
  - unread and attention state.
- Keep hidden archived lineage rows hidden as standalone rows while still using them as sidebar-state references.
- Avoid misclassifying ordinary collapsed lineage segments as child sessions.
- Update pin/unpin to mutate the cached `_allSessions` row and render immediately, then reconcile with the server.
- Add focused regression coverage for streaming-state detection, stale stream-id false-spinners/unreads, hidden archived lineage bubbling, collapsed-lineage child-badge avoidance, polling fallback alignment, timestamp bubbling, and immediate pin cache rendering.

## Why It Matters

Users type into the visible conversation row, not into an internal archived continuation row. If the internal row is the one streaming, the sidebar should still show the visible conversation as running and keep it sorted with current activity. This preserves the mental model that the sidebar represents conversations, not every hidden implementation segment.

## Verification

- `node -c static/sessions.js`
- `python -m pytest tests/test_issue2454_active_session_spinner.py tests/test_issue856_background_completion_unread.py tests/test_session_lineage_collapse.py tests/test_sidebar_pin_immediate_reorder.py -q`
  - Result: `74 passed, 1 warning`
- Delegated implementation review and PR review found collapsed-lineage, stale-stream, and persisted-observed-stream edge cases; fixed before this update.
- Delegated sanitization review: `PASS_WITH_NITS`; the noted local absolute path in the PR body was removed before update.
- Local diff sanitization scan found no local paths, hostnames, ports, session IDs, usernames, emails, tokens, or private environment details.

## Risks / Follow-ups

- The change is intentionally limited to sidebar-derived render state. It does not mutate session archival state or lineage metadata on disk.
- Full local test runner was not used for final verification because focused sidebar coverage exercises the touched behavior; CI should provide the full matrix.

## Contract Routing

Task type: sidebar/session metadata behavior fix.

Touched areas:
- `static/sessions.js`
- sidebar lineage collapse / child-session decoration
- streaming completion polling fallback
- focused sidebar regression tests

Relevant public docs:
- `AGENTS.md`
- `CONTRIBUTING.md`
- `docs/CONTRACTS.md`
- `docs/UIUX-GUIDE.md`

State layer / invariant:
- Sidebar render-derived state only. Hidden archived lineage rows remain hidden, but their runtime state may bubble to the visible conversation row.

## Model Used

AI-assisted implementation and review using OpenAI GPT-5.5 via Hermes Agent, with terminal/file tools and two read-only delegate review passes.

### CODE REVIEW

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Thanks @santastabber — the bubbling mechanics here are genuinely well built, and I verified the two clean parts: the **ordering change is correct with no regression** (`_sidebar_activity_at` is recomputed each render via `cleanSidebarRow` deleting it at rebuild start, falls back to `last_message_at` for normal rows — no cross-render drift), and there's **no render-derived state leak** across session switches (all `_child_session_*` decoration is cleared per rebuild). Pin/unpin optimistic-then-reconcile is sound. I ran the full gate (Codex + Opus); Codex passed it, but Opus reproduced a real regression on the exact surface this touches, so I need one fix before shipping.

### (must-fix) Broadening `_isSessionEffectivelyStreaming` to trust raw `active_stream_id` causes a false spinner + false unread

The change makes `_isSessionEffectivelyStreaming` treat a row as streaming when it has an `active_stream_id`, **without reconciling that against the live stream registry**. But a dead-but-uncleared `active_stream_id` is routine after a restart or interrupted turn — that's precisely why the server computes `is_streaming = (active_stream_id ∈ live-set)` and why `_isSessionLocallyStreaming` was gated to the active session (the #2066 INFLIGHT fix). Consequences:

1. **False spinner:** a session `{is_streaming:false, active_stream_id:"dead-stream"}` now shows the running spinner on its visible row — and because the definition changed globally, this fires for *every* row, not just hidden lineage children (wider blast radius than the PR title).
2. **False unread (the #856 surface):** `_markPollingCompletionUnreadTransitions` records the stale row as observed-streaming and persists it to localStorage (`_rememberObservedStreamingSession`). When server repair later clears `active_stream_id`, `completedPersistedObservedStream = observedStreaming && !isStreaming` fires → `_markSessionCompletionUnread(sid)` even though the session produced no output. Because it's in localStorage, the false unread **survives reload/bfcache**.
3. **Multiple hidden children OR together** (`bubbleSidebarState`), so a single stale/dead-stream child poisons the parent row.

The updated #856/#2454 tests relax the old exact-definition assertion to the broader set and add a happy-path `active_stream_id → streaming:true` test, but **none exercise the `active_stream_id` set + `is_streaming` false case** — so they codify the regression instead of guarding against it.

**Fix (small, localized):**
- Drop the raw `active_stream_id` term from `_isSessionEffectivelyStreaming`. The server already folds the *live* `active_stream_id` into `is_streaming`, so the term is redundant when correct and harmful when stale. Keep `is_streaming` + the pending-user-message markers + the active-session-local check.
- `pending_started_at` is also used unreconciled — either confirm the backend clears it on abnormal termination, or gate it the same way.
- Add a regression test for the stale case: `{is_streaming:false, active_stream_id:"dead"}` must NOT be effectively-streaming (and must not produce a completion-unread transition), at both the row level and bubbled to the parent.

Everything else (ordering, pin/unpin, cleanup, the bubbling itself) is fine — once the streaming definition stops trusting the unreconciled `active_stream_id`, I'm happy to re-gate and ship.

**Comment by santastabber** [comment]
Thanks for the review — updated in `75d6162f`.

Addressed the requested blocker and earlier hardening notes:

- Rebased cleanly onto current `origin/master` (`v0.51.509`).
- Removed raw `active_stream_id` and standalone `pending_started_at` from `_isSessionEffectivelyStreaming`; effective streaming now relies on server-confirmed `is_streaming`, explicit pending-message markers, or the active local `S.busy` path.
- Guarded the polling/unread transition so a stale/dead `active_stream_id` does not create a false unread marker.
- Added regressions for stale `active_stream_id` false-spinner / false-unread / parent-bubbling cases.
- Hardened hidden lineage bubbling to require archived reference rows, fall back through parent id when `_lineage_root_id` is absent, and use clearer timestamp extraction.
- Removed the accidental PR-body scratch file from the branch and updated the PR body.

Local focused verification before push:

```text
node -c static/sessions.js
python -m pytest tests/test_issue2454_active_session_spinner.py tests/test_issue856_background_completion_unread.py tests/test_session_lineage_collapse.py tests/test_sidebar_pin_immediate_reorder.py -q
# 73 passed, 1 warning
```

PR readback showed the branch is now mergeable and CI has started.

**Comment by nesquena-hermes** [comment]
## Summary

Pulled the branch into a read-only worktree at HEAD `75d6162f` (the rebased single commit) and diffed against `origin/master`. The lineage-bubbling work and the pin-cache immediate-render path look solid. But the `completedPersistedObservedStream` issue Greptile flagged is **still live at HEAD** — and it's worse than a dead branch: this PR actively *regresses* a line that was correct on master.

## Code reference — this is a regression, not a pre-existing dead path

The diff at `static/sessions.js` changes the term (diff hunk `@@ -614,7 +620,7 @@`):

```diff
-    const completedPersistedObservedStream = Boolean(observedStreaming && !isStreaming);
+    const completedPersistedObservedStream = Boolean(observedStreaming && hasServerRunSignal && !isStreaming);
```

On `origin/master` the condition was `observedStreaming && !isStreaming` — correct. This PR adds `&& hasServerRunSignal`, and that turns it into a logical contradiction that can never be true. Tracing the two helpers at HEAD:

```js
// :370
function _isSessionEffectivelyStreaming(s) {
  return Boolean(s && (s.is_streaming || _hasPendingUserMessageSignal(s) || _isSessionLocallyStreaming(s)));
}
// :612
const hasServerRunSignal = Boolean(s.is_streaming || _hasPendingUserMessageSignal(s));
```

So `isStreaming === (hasServerRunSignal || _isSessionLocallyStreaming(s))`, which means **`hasServerRunSignal === true` always implies `isStreaming === true`**. Therefore `hasServerRunSignal && !isStreaming` is unsatisfiable, and `completedPersistedObservedStream` is permanently `false` (`static/sessions.js:623`).

## Why it matters

`observedStreaming` is the only one of the three completion signals that survives a page refresh — it's read from `localStorage` (`SESSION_OBSERVED_STREAMING_KEY`, `_getSessionObservedStreaming():301`), whereas `wasStreaming` (`_sessionStreamingById`) and `previousSnapshot` (`_sessionListSnapshotById`) are in-memory and reset to empty on reload. So this term is exactly the path that marks a background session unread when:

1. the session was streaming before a refresh (persisted to localStorage),
2. the user reloads,
3. the first poll after reload shows the server has finished (`!isStreaming`) with no `message_count` delta the snapshot path would catch.

`completedObservedStream` can't cover that case (it needs `wasStreaming === true`, gone after refresh) and `completedWithNewMessages` needs a count/timestamp delta vs a snapshot that no longer exists. Net effect: this PR silently breaks post-refresh unread marking that works on master today.

## Recommendation

Revert that one term to its master form (`static/sessions.js:623`):

```js
const completedPersistedObservedStream = Boolean(observedStreaming && !isStreaming);
```

`!isStreaming` already implies `!hasServerRunSignal`, so the dropped clause was strictly redundant in the satisfiable direction and harmful in the other. The `_forgetObservedStreamingSession(sid)` call at `:636` still fires right after, so it marks unread exactly once and then clears the localStorage entry — no repeat-fire risk.

## Test plan

The current suite passes because none of the four files exercise the post-refresh runtime path (Greptile's note on `test_issue856_background_completion_unread.py` is right — the new assertions check source strings, not behavior). Recommend a regression that: seeds `localStorage[SESSION_OBSERVED_STREAMING_KEY]` for a sid, clears `_sessionStreamingById` / `_sessionListSnapshotById` to simulate a reload, feeds a poll row with `is_streaming:false` and an unchanged `message_count`, and asserts `_markSessionCompletionUnread` is called once. That test fails at HEAD and passes after the one-line revert.

The rest of the PR (hidden archived lineage bubbling, the `_isServerIdleSessionRow` stale-stream guard at `:382`, immediate pin cache mutation) reads correctly to me — this single line is the blocker.

**Comment by santastabber** [comment]
Updated again in response to the latest review.

What changed:
- Rebased onto current `origin/master` (`v0.51.513`).
- Restored `completedPersistedObservedStream` to the master behavior:
  ```js
  Boolean(observedStreaming && !isStreaming)
  ```
- Added a runtime regression for the post-refresh persisted observed-stream path: seed observed streaming state, clear in-memory tracking/snapshots, feed an idle poll row, and assert unread is marked once and the observed marker is cleared.
- Preserved current `CHANGELOG.md` release entries; the diff now only adds the PR’s Unreleased note.
- Re-ran delegate verification on the exact local fix; verdict: `PASS`.

Verification before push:

```text
node -c static/sessions.js
python -m pytest tests/test_issue2454_active_session_spinner.py tests/test_issue856_background_completion_unread.py tests/test_session_lineage_collapse.py tests/test_sidebar_pin_immediate_reorder.py -q
# 74 passed, 1 warning
```

Sanitization scan stayed clean.

**Comment by nesquena-hermes** [comment]
## Follow-up review — regression resolved at HEAD `509e312c`

Re-pulled the branch into a read-only worktree (now at `509e312c`, rebased onto `v0.51.513`) and re-diffed `static/sessions.js` against `origin/master`. The `completedPersistedObservedStream` regression I flagged last round is fixed — the term is back to the master definition.

## Code reference

`static/sessions.js:623` now reads exactly what master has:

```js
const completedPersistedObservedStream = Boolean(observedStreaming && !isStreaming);
```

`git show origin/master:static/sessions.js` has the identical expression at its line 617, so the active-streaming-only narrowing that broke the persisted-observed path is gone. The new `completedObservedStream` guard is the right shape too (`static/sessions.js:612-614`):

```js
const hasServerRunSignal=Boolean(s.is_streaming||_hasPendingUserMessageSignal(s));
const canMarkCompletedStream=Boolean(hasServerRunSignal||previousSnapshot||observedStreaming);
const completedObservedStream = canMarkCompletedStream&&wasStreaming === true && !isStreaming;
```

This only fires the polling-fallback unread on a real observed→stopped transition that had a server run signal or a prior snapshot, so a stale `wasStreaming===true` with no corroboration can't synthesize a false unread.

## On the effective-streaming change

`_isSessionEffectivelyStreaming` (`static/sessions.js:370-381`) now folds in `_hasPendingUserMessageSignal(s)` alongside `s.is_streaming` and `_isSessionLocallyStreaming(s)`, and deliberately drops the raw `active_stream_id` / `pending_started_at` signals — which is what avoids the stale-`active_stream_id` false spinner. `_isServerIdleSessionRow` (`static/sessions.js:382`) correctly adds `!s.pending_started_at` to the idle predicate so the two stay consistent.

## Test plan / verification

The two new runtime regressions in `tests/test_issue856_background_completion_unread.py` cover both directions and are the right kind of test (they execute the actual extracted function bodies via `node -e`, not just source-text asserts):

- `test_polling_transition_ignores_stale_active_stream_id_without_server_streaming` — seeds `_sessionStreamingById['stale']=true`, feeds an idle row with a dead `active_stream_id`, asserts `{"unread": [], "observed": false}`.
- `test_polling_transition_marks_persisted_observed_stream_after_reload` — seeds observed streaming, clears in-memory tracking, feeds an idle poll row, asserts `{"unread": ["done"], "observed": {}, "streaming": false}`.

The second one is exactly the post-refresh persisted path the prior review was worried about, and it asserts the marker is consumed (`observed` empties) so it can't re-fire. CHANGELOG is also down to a single Unreleased entry now (`CHANGELOG.md:4-8`), resolving the earlier double-entry note. From a static read this looks ready; I haven't run the full suite on the worktree.

**Comment by santastabber** [comment]
Rebased again onto current `origin/master` (`v0.51.520`) after the base moved. No functional changes beyond preserving the existing PR diff on the new base.

Re-verified before push:

```text
node -c static/sessions.js
python -m pytest tests/test_issue2454_active_session_spinner.py tests/test_issue856_background_completion_unread.py tests/test_session_lineage_collapse.py tests/test_sidebar_pin_immediate_reorder.py -q
# 74 passed, 1 warning
```

Sanitization scan stayed clean, and `CHANGELOG.md` still only adds the single Unreleased note for this PR while preserving all release entries.

**Comment by santastabber** [comment]
The latest code/test matrix is green except for `browser-smoke`, which appears to be a transient resource-load smoke failure rather than a product regression.

Evidence:
- All pytest shards and lint passed on PR head `4ed8ecfa`.
- `browser-smoke` failed only on `/#sessions` with three generic console resource-load errors:
  ```text
  [/#sessions] console: Failed to load resource: net::ERR_FAILED
  ```
  `/` and `/#settings` both loaded cleanly.
- I reproduced `tests/browser_smoke.py` locally on the PR-head worktree with Playwright installed:
  ```text
  OK  / — no console errors
  OK  /#settings — no console errors
  OK  /#sessions — no console errors
  BROWSER SMOKE PASSED — all pages loaded with zero console errors
  ```
- Recent `master` browser-smoke runs are passing.

I attempted to rerun the failed job with `gh run rerun 27847385466 --failed`, but GitHub rejected it: `Must have admin rights to Repository`.

Could someone with permissions rerun the browser-smoke workflow/check?

**Comment by santastabber** [comment]
Did small change to force checks again. Green now

**Comment by nesquena** [comment]
@santastabber can you email me? I'd like to add you to our dev discord

**Comment by nesquena-hermes** [comment]
Shipped in v0.51.601 (Release VH) — thank you @santastabber. The re-push fixed the regression exactly: effective-streaming no longer trusts the raw active_stream_id, so a dead-but-uncleared stream can't trigger a false spinner or a false unread that survives reload, and your new regression test locks the stale-stream case. Codex SAFE, full suite 10183 green, verified live in prod.
